extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js

```diff
@@ -138,7 +139,20 @@ export class AmpAdXOriginIframeHandler {
         this.adResponsePromise_,
         'timeout waiting for ad response').catch(e => {
           this.noContent_();
-          user().warn('AMP-AD', e);
+          getAdCid(this.baseInstance_).then(cid => {
+            const tagType = this.baseInstance_.element.getAttribute('type');
+            const TAG = `AMP-AD-${tagType}`;
+            dev().error(TAG, JSON.stringify({
+              'm': e instanceof Error ? e.message : e,
+              'type': tagType,
+              'correlator': this.baseInstance_.win.ampAdPageCorrelator ||
+                  null,
+              'slotId': this.baseInstance_
+                  .element.getAttribute('data-amp-slot-index'),
+              'clientId': cid,
```

Comment:
I was afraid of that.  So let's consider alternatives.  The issue is that we'd like to be able to capture enough info to reconstruct (if possible) the ad request that generated the no-fill.  We'd like to be able to figure out whether there's a specific class of ads that is failing.

There's a creative-specific ID that is returned in the ad response, but we don't have access to it from outside the frame.  We were hoping to catch the inputs that are used to construct the ad request so that we can reconstruct it on the backend, if necessary.

Is there an existing mechanism to log to ad network-provided destinations?  amp-analytics, perhaps?  Can we wire it to fire in this kind of case?